[oneDAL] Glossary section (#134)
oneapi-src/oneAPI-spec: source/elements/oneDAL/source/glossary.rst

.. _glossary:

=========
Glossary
=========

Machine learning terms
======================

.. glossary::
    :sorted:

>     Attribute

Reviewer: We might have a 'main' term and then list its synonyms in the same line as Ekaterina suggested somewhere else

        A particular property or quality of a real object or an event.
        Has a defined type and domain.

    Classification
        A supervised machine-learning problem to assign categorical :term:`labels<Label>`

>         to :term:`feature vectors<Feature vector>` describing input objects.

Reviewer 2: * problem of assigning
* do you need "describing input objects" here? don't feature vectors always describe input objects?

    Dataset
        A collection of :term:`instances<Instance>`.

    Feature
        A pair of :term:`attribute` information and a corresponding value.

    Feature vector
        A vector that encodes information about single real object or an event.
        Contains at least one :term:`feature`.

    Inference

>         A process of obtaining outputs from the :term:`model` created on the :term:`training` stage

Author: Reformulated.

    Instance

>         A vector of at least one :term:`feature` value and zero or more :term:`label<Label>` values.

Reviewer 2: what about regression? you don't always have labels

Author: Labels are applicable for the regression task too. For unsupervised learning tasks, there are no labels present, that's why I wrote "zero or more".

Reviewer: do we use 'label' in the context of regression?

Reviewer 2: label in the context of regression sounds weird to me, mostly because in my understanding labels are categorical

        Each of :term:`labels<Label>` and :term:`features<Feature>` can encode different objects or events.

    Label

>         :term:`Feature` of some object or event which is present on the :term:`training` stage but

Reviewer 2: **at** the training stage

>         unavaliable on the :term:`inference` stage.

Reviewer 2: * it's not only unavailable, but it's something you predict at the inference stage -- I think it should be mentioned
* **at** the inference stage

Author: Reformulated.

>         unavaliable on the :term:`inference` stage.

Reviewer: When I train a model I can use the part of the available data set to do that. The rest of the data set incl labels can be used to test the model by predicting and comparing vs known label

Author: Yes, but you are talking about test or validation stages, not inference. Do you mean I should mention these stages too?

Reviewer: I mean that we formally use the inference stage to validate the model. I assume you are talking about inference stage where you use earlier unseen data. Probably 1) a better wording is required to describe what the 'label' is, 2) unavailable label is more about specifics/def of supervised learning, and that place would be better to discuss it

Reviewer 2: why are we describing what label is as "something that present at the stage A and unavailable at the stage B"? it is not a definition of a label.
I'm still confused how labels are present in regression, but if we are talking about labels in classification, I'd highlight that 1) labels are categorical 2) we talk about labels when we discuss the response (outcome, target variable), not the independent variables 3) it's relevant to supervised learning, as Andrey mentioned above 4) you can mention how labels are used at the training stage vs inference 5) might be a good idea to give examples

>     Model

Reviewer: general question: do we rely on some literature that introduces the terms or define our self

Author: I formulate it by myself, but with accordance to external resources. I use [this](https://ml-cheatsheet.readthedocs.io/en/latest/glossary.html) and [this](https://developers.google.com/machine-learning/glossary) links and Peter Flach's book to check the terms.

Reviewer: option to think: if there is a well recognized a book/source with thedefs we might rely on it and then refer that we use this source as the base for our ml glossary in the volumes necessary for the lib

Author: Yes, it might be good and probably will give a "weight" to our terminology in the reader's eyes. 
But I really don't know, can we call the resources which I list somewhere below as a "well recognized"? How to measure it? I mostly rely on the **number** of resources and their **difference** to formulate more or less "objective" definitions.
Another thing is that most of the books have very long, "continuous" terms, which need to be reformulated shortly. Can we refer to a book if we mostly reformulated the term which is described in it?
 @andreynv @rlnx can you please name some well-recognized sources which I should also consider?

Reviewer: Mentioned one somewhere in the other comment and let you know, if I have another

        A result of :term:`training` machine-learning algorithm that stores information necessary to run
        :term:`inference` on a new :term:`dataset`.

    Observation
        A synonym for :term:`instance`.

    Prediction
        - A synonym for :term:`inference`.
        - An output of the :term:`trained<Training>` :term:`model` when provided with an input :term:`instance`.

>     Response

Author: Cannot see why. Labels are for both discrete and continuous features. Are you asking to delete an example from this term?

>     Response

Reviewer 3: Is there any book/article that uses the term `label` for regression problems? I suggest to stick to more adopted terminology.

        A value of a single :term:`label`. For example, response may be a flag
        indicating that an observation belongs to a particular class.

    Sample
        A synonym for :term:`instance`.

>     Training

Author: Changed to "A search process of the best model" - is it more accurate?

>     Training

Reviewer 2: It's grammatically incorrect (search **for**)
But I also don't like this definition because it doesn't say anything about how you determine which model is the best. What exactly are you searching for? What do you have at the beginning of that search (e.g. you need a training set and a target variable).
I'd mention parametric and non-parametric methods here (q: should there be entries for them in the glossary?)

>     Training

Author: Reformulated, please look. It is very hard to bring any details about this term because it is very common. I cannot even say that it is an optimization process, that's why I used the term "search process".

        A search process of the best parameter values of the :term:`model`.

>     Variable

Reviewer 3: Alias for the attribute/feature

        A synonym for an :term:`attribute`.

|dal_short_name| library terms
==============================

.. glossary::
    :sorted:

    Batch Mode
        The computation mode for an algorithm in |dal_short_name|, where all the data needed for
        computation is available at the start and fits the memory of the device on which the computations are performed.

    Immutability

>         The object is immutable when it is not possible to change its state after it is constructed.

Reviewer 2: I'd change to: 
The object is immutable if it is not possible to change its state after creation.

    Online Mode
        The computation mode for an algorithm in |dal_short_name|, where the data needed for computation
        becomes available in parts over time.

>     Workload

Reviewer: in addition to those terms, we would need terms related to features and necessary to describe data management: continuous feature, data layout, ....

Author: Yes, I've added some of the terms. Now, there is a question about *data layout* - should we introduce this term and entity or not. I think we need to discuss metadata API in details, I will prepare for it. Now I don't introduce this term.

Reviewer: The specifics here is that while we can use 'data layout', 'data format' etc in the specific dal sense, the real sense of the terms is generally different. The idea-to make sure we have defs aligned with community's understanding to avoid confusion; moreover it would be used in apis somewhere

        A task of applying a |dal_short_name| algorithm to a dataset.

Other
=====

> TODO: this section looks like a general one for the whole oneAPI spec.

Reviewer: yes, it should be in the common glossary

Author: I mean, this section should be deleted when common glossary appears. Now I don't want to remove it because we have a references on the spec to these terms, which will be broken after removal. @rlnx @andreynv @outoftardis what is your opinion, should I delete it now?

Reviewer: I'd have it here until a common mechanism for dictionaries/bibliographies is agreed and implemented on oneapi level

Reviewer 2: I agree with @andreynv. But don't call this section "Other".  oneAPI terms, maybe? or common oneAPI terms?

.. glossary::
    :sorted:

    API
        Application Programming Interface

    DPC++
        Data Parallel C++ (DPC++) is a high-level language designed for data parallel programming productivity.
        DPC++ is based on :term:`SYCL* <sycl>` from the Khronos* Group to support data parallelism and heterogeneous programming.

    Host/Device
        OpenCL [OpenCLSpec]_ refers to CPU that controls the connected GPU executing kernels.

    JIT
        Just in Time Compilation --- compilation during execution of a program.

    Kernel
        Code written in OpenCL [OpenCLSpec]_ or :term:`SYCL` and executed on a GPU device.

    SPIR-V
        Standard Portable Intermediate Representation - V is a language for intermediate representation of compute kernels.

    SYCL
        SYCL(TM) [SYCLSpec]_ --- high-level programming model for OpenCL(TM) that enables code for heterogeneous

>         processors to be written in a "single-source" style using completely standard C++.

Reviewer 3: Please take a look at #122, those terms are going to be common for the entire spec

Author: Please look at the description of this PR. I mentioned that this section should be deleted once it moved to the top level.

.. _glossary:

=========
Glossary
=========

Machine learning terms
======================

.. glossary::
    :sorted:

    Categorical feature

>         A :term:`feature` with a discrete set of possible values without ordering.

Reviewer 2: I think it's finite, not discrete

Author: I meant (discrete data)[https://stats.stackexchange.com/questions/206/what-is-the-difference-between-discrete-data-and-continuous-data]. Maybe, it is better to reformulate it like this?
"A feature with a discrete domain without ordering between values"

Reviewer: 'discrete, continuous, ordinal, categorical' (and Mikhail A also suggested interval) were aligned with http://dmg.org/pmml/v4-1/DataDictionary.html in the lib

        Synonyms: nominal feature.

    Classification
        A :term:`supervised machine-learning problem<Supervised learning>` of assigning
        discrete :term:`labels<Label>` (:term:`categorical<Categorical feature>` or :term:`ordinal<Ordinal feature>`)
        to :term:`feature vectors<Feature vector>`.

>     Continuous feature

Author: Its debatable how to name such features. Some sources name it "numerical", some "continuous" or "quantitative". For example, Peter Flach (p.304):
> <...> we can distinguish three main kinds of feature: those with a meaningful numerical scale, those without scale but with an ordering, and those without either. We will call features of the first type quantitative <...> (another term in common use is 'continuous').

>     Continuous feature

Author: @rlnx @andreynv what is your opinion on how we should primarly name it? Three of my sources name it differently - Peter Flach calls it "quantitative", Jure Leskovec et all call it "numerical", Google ML glossary - "continuous". I chose the last one because it has been already used in DAAL.

>     Continuous feature

Reviewer: responded somewhere else. If we see different terms, we have freedom to stick to one of them

        A :term:`feature` with values in a domain of real numbers.

        Synonyms: quantitative feature.

    Dataset
        A collection of :term:`observations<Observation>`.

>     Feature

Reviewer 2: makes sense to link to the features of different types

and a general q: do we maybe want to list related terms? If something that might be relevant for the reader but not specifically appears in the definition

Author: I would not want to do this, because the glossary transforms into an ML guide. In my opinion, the terms should be short, like in vocabulary. The reader must be referenced to the **direct** term he wants to clarify in the spec. And this term can refer to other terms. 

@rlnx @andreynv what is your opinion?

Reviewer: the quick example from the top of my mind. During the design of dbscan the terminology that came from the paper was 'points' or so which means observation of feature vector for the lib. To make sure that the dbscan is described in properly, we introduced that synonym.  to simplify, we might have the same term and rely on it across the specs even if it would deviate from the terminology in the paper (for 'dbscan' we might refer to the paper and mention that it uses 'point' instead of 'observation')

        A particular property or quality of a real object or an event.
        Has a defined type and domain.

        Synonyms: attribute, variable

    Feature value
        A single point in the domain of the :term:`feature`.

    Feature vector
        A vector that encodes information about single real object or an event.
        Contains at least one :term:`feature value`.

    Inference
        A process of applying a :term:`trained<Training>` :term:`model` to the :term:`inference set` with a goal to
        predict :term:`responses<Response>` of the :term:`labels<Label>`.

        Synonyms: prediction.

    Inference set
        An unlabeled :term:`dataset` used at the :term:`inference` stage to make an outputs from the :term:`model`.

    Label
        :term:`Feature` of some object or event which is unavailable at the :term:`inference` stage

>         and needs to be predicted.

Reviewer 3: It's too debatable to use the `label` for continuous variables, [the same question here](https://github.com/oneapi-src/oneAPI-spec/pull/134/files#r424726360). Maybe introduce more liberal term like `target variable`?

Author: @rlnx @andreynv @outoftardis guys, I am not inventing my own ML terms. Here are the sources which have this term:
- Peter Flach, Machine Learning. The Art and Science of Algorithms that Make Sence of Data, Cambridge University Press, 2012. page 50.
> We furthermore distinguish between the label space `L` and the output space `Y`. The label space is used in supervised learning to label examples. <...> The most commonly encountered machine learning scenario is where the label space coincides with the output space. That is, `Y=L` <...>. This scenario covers both classification and regression. In cases where the label space and the output space differ, this usually serves the purpose of learning a model that outputs more than just a label <...>.
- Jure Leskovec, Anand Rajaraman, Jeffrey D. Ullman, Mining of Massive Datasets, 2014. page 440
> A training set consists of a set of pairs `(x, y)` <...> where `y` is a label, the classification value for `x`. <...> The type of `y` is in principle arbitrary, but there are several common and important cases. `y` is real number. In this case, the ML problem is called regression. `y` is a boolean value <...>. In this class the problem is binary classification. <...>
- Google`s ML glossary: [label](https://developers.google.com/machine-learning/glossary#label)
- [open-source ML glossary](https://ml-cheatsheet.readthedocs.io/en/latest/glossary.html#glossary-observation)
> Label
The “answer” portion of an observation in supervised learning.
- Direct [question](https://stackoverflow.com/questions/40898019/what-is-the-difference-between-a-feature-and-a-label) about the label on stack overflow.
>Briefly, feature is input; label is output. This applies to both classification and regression problems.

If you think that the term "target variable" or some other term is better, I can change for it.

Reviewer: This is the mess with terminology.
This book was suggested by the lead of ML group at TMG, http://faculty.marshall.usc.edu/gareth-james/ISL/.  It uses 'quantitative response', response in case of regression and 'binary qualitative response' for classification. I'd clearly differentiate this dependent variable for regression and classification. Response or target variable or similar for regression and label for classification. If the desire to connect them, say about response/target variable in generic sense, then specify quantitative and connect to regression with the right term, specify qualitative and connect to label

Author: Used `response` term from this book to reference output variables for both classification/regression, changed the `label` term point only to discrete features.

.. _glossary:

=========
Glossary
=========

Machine learning terms
======================

.. glossary::
    :sorted:

    Categorical feature
        A :term:`feature` with a discrete domain without ordering between values.

>         Synonyms: nominal feature.

Reviewer: here we specify synonyms in one style and somewhere else in another style?

Author: I specify synonyms consistently across this document. @outoftardis, is it ok to indicate synonyms as I do? Or we have some other approved/used style.

Reviewer 2: we don't have any other style, yours is ok, but I suggest these changes: 
- remove the full stop
- make the word **Synonyms** bold
- use **Synonym** if there is only one

The same goes for examples/example.

Author: Done

        discrete :term:`labels<Label>` (:term:`categorical<Categorical feature>` or :term:`ordinal<Ordinal feature>`)
        to :term:`feature vectors<Feature vector>`.

    Clustering
        An :term:`unsupervised machine-learning problem<Unsupervised learning>` of grouping
        :term:`feature vectors<Feature vector>` onto (usually
        :term:`categorical<Categorical feature>`) groups without prior information on them.

    Continuous feature
        A :term:`feature` with values in a domain of real numbers.

        Synonyms: quantitative feature, numerical feature.

    Dataset
        A collection of :term:`observations<Observation>`.

    Feature
        A particular property or quality of a real object or an event.
        Has a defined type and domain.

        Synonyms: attribute, variable

    Feature value
        A single point in the domain of the :term:`feature`.

    Feature vector
        A vector that encodes information about single real object or an event.
        Contains at least one :term:`feature value`.

    Inference
        A process of applying a :term:`trained<Training>` :term:`model` to the :term:`inference set` with a goal to

>         make outputs based on input :term:`feature vectors<Feature vector>`.

Reviewer: make outputs may not be clean

Author: Reformulated

        Synonyms: prediction.

    Inference set
        An unlabeled :term:`dataset` used at the :term:`inference` stage to make an outputs from the :term:`model`.

    Label
        A :term:`feature` of some object or event which value is target to be predicted in :term:`supervised learning` task.
        On the :term:`training` stage, the model which finds the best dependency between input :term:`feature vector` and
        :term:`labels<Label>` is selected. :term:`Label` is one of outputs can be made by :term:`model`
        on the :term:`inference` stage.

    Model
        A result of :term:`training` machine-learning algorithm that stores information necessary to run
        :term:`inference` on a new :term:`dataset`.

    Observation
        A vector of at least one :term:`feature` value and zero or more :term:`label<Label>` values.

>         Each of :term:`labels<Label>` and :term:`features<Feature>` can encode different objects or events.

Reviewer: this label should be stabilized when we close this discussion above

        Synonyms: instance, sample.

    Ordinal feature
        A :term:`feature` with a discrete set of possible values and some ordering between them.

    Regression
        A :term:`supervised machine-learning problem<Supervised learning>` of assigning
        :term:`continuous<Continuous feature>` :term:`labels<Label>` to

>         :term:`feature vectors<Feature vector>`.

Reviewer: the definition is still incomplete

Author: Changed `label` to `response`. Cannot see why it is incomplete. Even in the source mentioned by you, the regression task is described similar to my definition:
>We tend to refer to problems with a quantitative response as regression problems

    Response
        A value of a single :term:`label`. For example, response may be a flag
        indicating that an observation belongs to a particular class.

    Supervised learning
        :term:`Training` process with :term:`labeled<Label>` :term:`training set`.

>         The best :term:`model` of dependencies between input :term:`feature vector`

Reviewer: best may sound confusing ; need a more neutral term

Author: Reformulated

        and :term:`labels<Label>` is selected.

    Training
        A process of creating a best :term:`model` (in terms of some quality criteria) based on
        information extracted from a :term:`training set`.

    Training set
        A :term:`dataset` used at the :term:`training` stage to create a :term:`model`.

    Unsupervised learning
        :term:`Training` process where :term:`training set` does not contain information

>         of dependencies between :term:`feature vectors<Feature vector>` and outputs need

Reviewer: information on dependencies

        to be produced by the :term:`model`.

|dal_short_name| terms
======================

.. glossary::
    :sorted:

    Batch Mode
        The computation mode for an algorithm in |dal_short_name|, where all the data needed for
        computation is available at the start and fits the memory of the device on which the computations are performed.

    Contiguous data
        A special property of the :term:`data format`: the data are stored as one contiguous memory block.

    Data format
        The representation of internal structure of the data: this information is enough to efficiently access them.

    Data type

>         An integral or floating-point type which is used to physically store the data.

Reviewer: what term is used by c++ standard, integral, basic, scalar?

Author: - *character types` for `char`, `unsigned char`, ...
- *integer types* for `int`, `long`, `short` ...
- *integral types* for `bool` + character types + integer types

    Homogeneous data

>         A special property of the :term:`data format`: all the :term:`features<Feature>` have the same type and domain.

Reviewer: that's interesting: both features are double, but one belongs to (0,1) and another - to (2,3). Is it homogeneous or not?

Author: In my point of view, it is not (and this is mentioned in the definision of the term "have the same type and **domain**"). If data are homogeneous, then we can act with all the features similarly, we don't need to check the details about the types of each feature - it's a verbose simplification. If we allow some exceptions (like with domain), it may be unclear. I think that one needs to use homogeneous data if and only if it is claimed that all the features can be handled similarly.

Reviewer: We potentially limit the scope of 'homogeneous data', only the data the belong to the same domain would be considered homogeneous. To me, the important aspect to keep in mind to differentiate between homo-/heterogeneous data - a set of ops we can do on the type: say, if we can apply multiplication, summation, comparison, etc to the data represented with the same data type, they are about the same, and domain is secondary and serves for the next level of checks on validity of the data or so during the compute stage or in error handling

Author: Yes, I agree with you. It was my misunderstanding that domain defines the available operations too.

    Immutability
        The object is immutable if it is not possible to change its state after creation.

    Online Mode
        The computation mode for an algorithm in |dal_short_name|, where the data needed for computation
        becomes available in parts over time.

    Reference-counted object
        A |dal_short_name| object which stores the number of references to the unique implementation.

>         Such object has a lightweight copy constructor and assignment operator, which simply increases the number of references.

Reviewer: it requires def of copy constructor/assignment operator - is it possible to simplify?
'simply' - remove

Author: Reformulated with C++ terminology - copy-assignable and copy-constructible. Thanks!

.. _glossary:

=========
Glossary
=========

Machine learning terms
======================

.. glossary::
    :sorted:

    Categorical feature
        A :term:`feature` with a discrete domain. Can be :term:`nominal <Nominal
        feature>` or :term:`ordinal <Ordinal feature>`.

        **Synonyms:** discrete feature, qualitative feature

    Nominal feature
        A :term:`categorical feature` without ordering between values. Only
        equality operation is defined for nominal features.

        **Examples:** a person's gender, color of a car

    Classification
        A :term:`supervised machine learning problem <Supervised learning>` of
        assigning :term:`labels <Label>` to :term:`feature vectors <Feature
        vector>`.

        **Examples:** predict what type of object is on the picture (a dog or a cat?),
        predict whether or not an email is spam

    Clustering
        An :term:`unsupervised machine learning problem <Unsupervised learning>`
        of grouping :term:`feature vectors <Feature vector>` into groups,

>         usually :term:`categorical <Categorical feature>`,

Reviewer 3: What does "categorical groups" mean? The attempt to logically connect groups and categorical features seems irrelevant here

Reviewer 3: "grouping into groups" sounds odd

Reviewer 3: An unsupervised machine learning problem of grouping a set of feature vectors in a way that observations in the same group (called a cluster) are more similar to each other than to those in other clusters.

Author: Rephrased a bit

        Each of them can encode different objects or events.

        Synonyms: instance, sample.

    Ordinal feature
        A :term:`categorical feature` with defined operations of equality an
        ordering between values.

        Examples: student's grade.

    Ratio feature
        A :term:`continuous feature` with defined operations of equality,
        comparison, addition, substraction, multiplication and division.
        Has a zero value element which means the absence of value.

        Examples: the height of a tower.

    Regression
        A :term:`supervised machine-learning problem<Supervised learning>` of
        assigning :term:`continuous<Continuous feature>`
        :term:`responses<Response>` to :term:`feature vectors<Feature vector>`.

        Examples: prediction of the temperature based on weather conditions.

    Response
        A property of some real object or event which value is target to be
        predicted in the :term:`supervised learning` task. While the
        :term:`feature` is an input in the machine learning task, the response
        is one of the outputs can be made by the :term:`model` on the

>         :term:`inference` stage.

Author: Reformulated

        values based on input :term:`feature vectors<Feature vector>`.

        **Synonym:** prediction

    Inference set
        A :term:`dataset` (usually without :term:`responses<Response>` present)
        used at the :term:`inference` stage.

    Interval feature
        A :term:`continuous feature` which values can be compared, added or
        subtracted, but cannot be multiplied or divided.

        **Examples:** a timeframe scale, a temperature in Celcius or Fahrenheit

    Label
        A :term:`response` with :term:`categorical<Categorical feature>` or
        :term:`ordinal<Ordinal feature>` values. This is an output in
        :term:`classification` and :term:`clustering` tasks.

        **Example:** the spam-detection task has a binary label indicating whether
        the email is spam or not

    Model
        An entiry which stores an information necessary to run :term:`inference`
        on a new :term:`dataset`. Typically a result of :term:`training`
        process.

        **Example:** in linear regression algorithm, the model contains weight
        values for each input feature and single bias value

    Observation

>         A :term:`feature vector` and zero or more :term:`responses<Response>`.
>         Each of them can encode different objects or events.

Reviewer 2: each of them - each of what?

```suggestion
        A :term:`feature vector` and zero or more :term:`responses <Response>`.
        Each of them can encode different objects or events.
```

Author: Second statement removed. Detailed description is written for feature vector.

        **Synonyms:** instance, sample

    Ordinal feature

>         A :term:`categorical feature` with defined operations of equality an
>         ordering between values.

Reviewer 2: ```suggestion
        A :term:`categorical feature` with defined operations of equality and
        ordering between values.
```

        **Example:** student's grade

    Ratio feature

>         A :term:`continuous feature` with defined operations of equality,
>         comparison, addition, substraction, multiplication and division.
>         Has a zero value element which means the absence of value.

Reviewer 2: ```suggestion
        A :term:`continuous feature` with defined operations of equality,
        comparison, addition, subtraction, multiplication, and division.
        Zero value element means the absence of any value.
```

        **Example:** the height of a tower

    Regression
        A :term:`supervised machine-learning problem<Supervised learning>` of
        assigning :term:`continuous<Continuous feature>`
        :term:`responses<Response>` to :term:`feature vectors<Feature vector>`.

        **Example:** prediction of the temperature based on weather conditions

    Response
        A property of some real object or event which value is target to be
        predicted in the :term:`supervised learning` task. While the
        :term:`feature` is an input in the machine learning task, the response
        is one of the outputs can be made by the :term:`model` on the
        :term:`inference` stage.

        **Synonym:** dependent variable

    Supervised learning

>         :term:`Training` process with a :term:`dataset` containing information
>         on dependencies between :term:`features<Feature>` and
>         :term:`responses<Response>`. The goal is to get a :term:`model` of
>         dependencies between input :term:`feature vector` and
>         :term:`responses<Response>`.

Reviewer 2: ```suggestion
        :term:`Training` process that uses a :term:`dataset` with information
        about dependencies between :term:`features <Feature>` and
        :term:`responses <Response>`. The goal is to get a :term:`model` of
        dependencies between input :term:`feature vector` and
        :term:`responses <Response>`.
```

Also note that here the reference goes to `dataset`, while in unsupervised learning you refer `training set`.

    Training
        A process of creating a :term:`model` based on information extracted
        from a :term:`training set`. Resulting :term:`model` is selected in
        accordance with some quality criteria.

    Training set
        A :term:`dataset` used at the :term:`training` stage to create a
        :term:`model`.

    Unsupervised learning

>         :term:`Training` process where :term:`training set` does not contain
>         information on dependencies between :term:`feature vectors<Feature
>         vector>` and :term:`responses<Response>`. The goal is to get a
>         :term:`model` matching :term:`feature vectors<Feature vector>` into
>         groups.

Reviewer 2: Not sure about dependencies. We don't have responses in unsupervised learning, right?
My suggestion:
```suggestion
        :term:`Training` process that uses a :term:`training set` with no
        :term:`responses <Response>`. The goal is to get a
        :term:`model` that allocates :term:`feature vectors <Feature vector>`  into
        groups.
```

Author: Good question about responses. At the training stage yes, we do not have them. But how to refer the outputs of the unsupervised task made at the inference stage? In my opinion, response can be used for supervised (regression+classification) tasks and unsupervised tasks. Label can be used for classification and clustering(unsupervised) tasks.
@rlnx , @andreynv , what is your opinion, can we use `response` for unsupervised tasks?

Reviewer: Unsupervised learning is about the case when we have feature vectors without responses and want to find some hidden patterns, e.g. how feature vectors are 'similar'. How we represent the outcome of this learning is secondary topic, e.g., we numerate the groups of the similar feature vectors or we have a set of rules that indicate connection between individual items/features in the data (if I buy this thing, I'll highly likely buy that thing)

Author: Ok, thanks.

|dal_short_name| terms
======================

.. glossary::
    :sorted:

    Accessor
        An |dal_short_name| concept of an object, which provides an access in the

>         special :term:`data format` to the data of another object.

Reviewer 2: also not sure about "concept of an object". what about "concept for an object"?

    Batch Mode
        The computation mode for an algorithm in |dal_short_name|, where all the
        data needed for computation is available at the start and fits the
        memory of the device on which the computations are performed.

    Builder

>         An |dal_short_name| concept of an object which encapsulates the creation
>         process of another object and provides an ability to create it
>         iteratively.

Reviewer 2: also same comment about concept as for accessor

```suggestion
        A |dal_short_name| concept of an object that encapsulates the creation
        process of another object and enables its iterative creation.
```

    Contiguous data
        A special property of the :term:`data format`: the data are stored as

>         one contiguous memory block.

Author: Just because it is a single attribute of data format. It does not fully describe data format: data can be continuous, heterogeneous, stored in row-major layout with feature types (double, float, int). That's the full description of data format.

>         one contiguous memory block.

Reviewer 2: would this be correct?

```
Data that is stored as one contiguous memory block.
One of the characteristics of a :term:`data format`.
```

>         one contiguous memory block.

Author: Yes, but I'd like to write "data that are stored..." because data is a plural noun. Correct me please, if I wrong.

>         one contiguous memory block.

Reviewer 2: there's a controversy :) 
generally both `data are`/`data is` are ok as long as you use them consistently, but there's a tendency towards data as a singular mass noun (similar to information)
if you look at other places in this doc, you'll see data used as a singular

    Data format
        The representation of internal structure of the data.

        **Examples:** data can be stored in array-of-structures or
        compressed-sparse-row format

    Data layout
        A property :term:`data format` of :term:`contiguous data`. Describes an

>         order of elements inside contiguous data block.

Author: Reformulated.

        **Example:** row-major format, where elements are stored row by row

    Data type
        An integral or floating-point type which is used to physically store the
        data.

        **Examples:** int32_t, float, double

    Heterogeneous data
        A property of the :term:`data format`, where :term:`features<Feature>`
        have different :term:`data types<Data type>` and/or operations defined

>         between values.

Author: Reformulated.

        **Example:** A :term:`dataset` with 100
        :term:`observations<Observation>` of 3 :term:`interval features<Interval
        feature>`. The first two features have data type of float32, but the
        third one of float64

    Homogeneous data
        A special property of the :term:`data format`: all the types of the
        :term:`features<Feature>` are the same. This means that they have

>         similar :term:`data type` and defined set of available operations.

Author: Reformulated.

        **Example:** A :term:`dataset` with 100
        :term:`observations<Observation>` of 3 :term:`interval features<Interval
        feature>` of type float32

    Immutability
        The object is immutable if it is not possible to change its state after
        creation.

    Online Mode
        The computation mode for an algorithm in |dal_short_name|, where the
        data needed for computation becomes available in parts over time.

    Reference-counted object
        A copy-constructible and copy-assignable |dal_short_name| object which
        stores the number of references to the unique implementation. Has
        lightweight copy operations which increases the number of references.
        An implementation is automatically freed when the number of references

>         becomes equal to zero.

Author: > not clear to me what is meant here

Such objects share a unique implementation. This implementation stores the number of these objects. When you creating a new object referencing to implementation, this number increases. The object can be created by copy-constructor or assignment operator - "copy operations". Usually, they perform heavy and costly full-copy actions, but here the only thing we do is just incrementing one number.

>         becomes equal to zero.

Reviewer 2: ```
Both copy operations defined for this object are lightweight, 
which means that each time a new object is created, only the number of references is increased by one.
```

>         becomes equal to zero.

Author: Reformulated

    Categorical feature

>         A :term:`feature` with a discrete domain without ordering between values.

Reviewer: to double-check: can categorical feature be ordinal, thus, ordinal and nominal are two classes of categorical features.  Also, double-check with Mikhail A. on another interval(?) class of features

Author: Thanks, that was a mistake. Added terms for nominal, interval and ratio features

        Has a defined type and domain. In machine learning tasks, feature is
        considered as input variable independent from others.

        Synonyms: attribute, variable, input variable.

    Feature vector
        A vector that encodes information about single real object or an event.
        Contains at least one :term:`feature`.

        Examples: A rectangle can be described by two features: its width and height.

    Inference
        A process of applying a :term:`trained<Training>` :term:`model`
        to the :term:`dataset` with a goal to make predictions of
        :term:`responce` values based on input :term:`feature vectors<Feature vector>`.

        Synonyms: prediction.

    Inference set
        A :term:`dataset` (usually without :term:`responses<Response>` present)
        used at the :term:`inference` stage.

    Label
        A :term:`response` with :term:`categorical<Categorical feature>` or :term:`ordinal<Ordinal feature>`
        values. This is an output in :term:`classification` and :term:`clustering` tasks.

        Examples: the spam-detection task has a binary label indicating whether the email is spam or not.

    Model

>         A result of :term:`training` machine-learning algorithm that stores information necessary to run

Reviewer: by machine learning algorithm?

Author: Reformulated

        A :term:`dataset` (usually without :term:`responses<Response>` present)
        used at the :term:`inference` stage.

    Label
        A :term:`feature` of some object or event which value is target to be predicted in :term:`supervised learning` task.
        On the :term:`training` stage, the model which finds the best dependency between input :term:`feature vector` and
        :term:`labels<Label>` is selected. :term:`Label` is one of outputs can be made by :term:`model`
        on the :term:`inference` stage.
        A :term:`response` with :term:`categorical<Categorical feature>` or :term:`ordinal<Ordinal feature>`
        values. This is an output in :term:`classification` and :term:`clustering` tasks.

        Examples: the spam-detection task has a binary label indicating whether the email is spam or not.

    Model
        A result of :term:`training` machine-learning algorithm that stores information necessary to run
        :term:`inference` on a new :term:`dataset`.

        Examples: in linear regression algorithm, the model contains weight values for each input feature
        and single bias value.

    Observation

>         A vector of at least one :term:`feature` value and zero or more :term:`label<Label>` values.

Reviewer: Just a feature vector and zero or more...

Author: Reformulated

    Response

>         A property of some real object or event which value is target to be predicted in the

Reviewer: is target to be predicted - we need a help of Ekaterina to word it in a cleaner way

Author: @outoftardis , please, help :)

Reviewer 2: sure, I'm reviewing the changes now and will add there a suggestion how to rephrase it

        A special property of the :term:`data format`: the data are stored as one contiguous memory block.

    Data format

>         The representation of internal structure of the data: this information is enough to efficiently access them.

Reviewer: we can or can't introduce data layout but we need to understand differences between two defs

Author: Added `data layout` term to show a difference between them

        Examples: data can be stored in array-of-structures or compressed-sparse-row format.

    Data type

>         An integral or floating-point type which is used to physically store the data.

Reviewer: just represent or describe the data?

Author: I think this term can be mentioned for both reasons. When we are talking about metadata, we mind "to describe and represent".  When we are talking about some variable which has a particular data type, we mean "to represent".

Reviewer: type is used to store the data..but the data does not seem to be stored by type

Author: Reformulated

        Examples: int32_t, float, double.

>     Homogeneous data

Reviewer: We also need to define heterogeneous data

Author: Done

        becomes available in parts over time.

    Reference-counted object

>         A copy-constructible and copy-assignable |dal_short_name| object

Reviewer: reference to the source that contains def of copy-constructible?

Author: Hmm... this is defined by the C++ standard, which is not freely available. I can add a reference to cppreference.com, which is a (non-official) reference for all terms and entities defined by C++ standard.
But do we really need that? This is a spec for C++/DPC++ library, so we can assume that the reader is familiar with C++ terms.

Reviewer: okay, if the others think in the same direction

        The object is immutable if it is not possible to change its state after
        creation.

>     Metadata
>         An information about logical and physical structure of an object. All
>         possible combinations of metadata values present the full set of
>         possible objects of a given type. Metadata do not expose an information
>         which is not a part of type definition, e.g. implementation details.

Reviewer 2: ```suggestion
    Metadata
        Information about logical and physical structure of an object. All
        possible combinations of metadata values present the full set of
        possible objects of a given type. Metadata do not expose information
        that is not a part of a type definition, e.g. implementation details.
```
